File: Back-end/NotificationService/src/main/java/com/example/NotificationService/service/NotificationService.java
package com.example.NotificationService.service;

import com.example.NotificationService.domain.Notification;
import com.example.NotificationService.exception.NotificationNotFoundException;
import com.example.NotificationService.repository.NotificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public interface NotificationService
{
    // save new notification
    Notification saveNotification(Notification notification);


    // mark notification as read
    Notification markNotificationAsRead(Long notificationId) throws NotificationNotFoundException;


    // fetch user (recipient) notifications with sorting and pagging
    Page<Notification> fetchRecipientNotification(String recipient);


    // filter user notifications based on date with sorting and pagging
    Page<Notification> findNotificationByDate(String recipient, LocalDate date);


    // filter user notifications based on month and year with sorting and pagging
    Page<Notification> findNotificationByMonthAndYear(String recipient, int month, int year);


    // find user unread notification with sorting and pagging
    Page<Notification> findUserUnreadNotification(String recipient);


    // count user unread notifications
    long countUnreadNotifications(String recipient);


}

File: Front-end/kanban-board-project/src/Services/NotificationService.js
import axiosInstance from "./axiosConfig";

const baseURL = "http://localhost:8084/api/v1/notification";

// get user notifications
const getAllNotifications = async (username) => {
   const response = await axiosInstance.get(`${baseURL}/allNotifications?recipient=${username}`);
   return response.data;
};


// mark notification as read
const markNotificationAsRead = async (notificationId) => {
   const response = await axiosInstance.post(`${baseURL}/markNotificationAsRead/${notificationId}`);
   return response.data;
};


// filter notification by date
const filterNotificationByDate = async (username, date) => {
   const response = await axiosInstance.get(`${baseURL}/getNotificationByDate?recipient=${username}&date=${date}`);
   return response.data;
};


// filter notification by month and year
const filterNotificationByMonthAndYear = async (username,month,year) => {
     const response = await axiosInstance.get(`${baseURL}/getNotificationByMonthAndYear?recipient=${username}&month=${month}&year=${year}`);
     return response.data;
};


// get unread notification
const unreadNotification = async (username) => {
    const response = await axiosInstance.get(`${baseURL}/getUnreadNotification?recipient=${username}`);
    return response.data;
};


// count unread notification
const countUnreadNotification = async (username) => {
    const response = await axiosInstance.get(`${baseURL}/countUnreadNotification?recipient=${username}`);
    return response.data; 
};


export {getAllNotifications, markNotificationAsRead, filterNotificationByDate, filterNotificationByMonthAndYear, unreadNotification, countUnreadNotification};



